smartsim/_core/launcher/dragon/dragonBackend.py

```diff
@@ -0,0 +1,155 @@
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import functools
+import typing as t
+from threading import RLock
+
+# pylint: disable=import-error
+from dragon.native.process import Process, TemplateProcess
+from dragon.native.process_group import ProcessGroup
+
+from smartsim._core.schemas import (
+    DragonHandshakeRequest,
+    DragonHandshakeResponse,
+    DragonRequest,
+    DragonResponse,
+    DragonRunRequest,
+    DragonRunResponse,
+    DragonShutdownRequest,
+    DragonShutdownResponse,
+    DragonStopRequest,
+    DragonStopResponse,
+    DragonUpdateStatusRequest,
+    DragonUpdateStatusResponse,
+)
+from smartsim.status import (
+    STATUS_COMPLETED,
+    STATUS_FAILED,
+    STATUS_NEVER_STARTED,
+    STATUS_RUNNING,
+)
+
+# pylint: enable=import-error
+
+
+class DragonBackend:
+    """The DragonBackend class is the main interface between
+    SmartSim and Dragon. It is not intended to be user-facing,
+    and will only be called by the Dragon entry-point script or
+    by threads spawned by it.
+    """
+
+    def __init__(self) -> None:
+        self._proc_groups: t.Dict[str, t.Tuple[ProcessGroup, t.List[int]]] = {}
+        self._step_id_lock = RLock()
+        self._step_id = 0
+
+    def _get_new_id(self) -> str:
+        with self._step_id_lock:
+            self._step_id += 1
+            return str(self._step_id)
+
+    @functools.singledispatchmethod
+    # Deliberately suppressing errors so that overloads have the same signature
+    # pylint: disable-next=no-self-use
+    def process_request(self, request: DragonRequest) -> DragonResponse:
+        raise TypeError("Unsure how to process a `{type(request)}` request")
```

Comment:
f-string appears to miss an `f"..."`

Reply:
Done as part of https://github.com/al-rigazzi/SmartSim/pull/7